File: src/main/kotlin/utils/UrlUtils.kt
package utils

import io.ktor.client.HttpClient
import io.ktor.client.request.head
import io.ktor.client.statement.HttpResponse
import io.ktor.http.HttpHeaders
import io.ktor.http.HttpStatusCode
import io.ktor.http.Url
import io.ktor.http.fullPath
import network.Http
import schemas.manifest.InstallerManifest

/**
 * Tries to extract the architecture from a URL and returns it as an [InstallerManifest.Installer.Architecture] enum
 * value.
 *
 * @return an [InstallerManifest.Installer.Architecture] enum value if an architecture can be found, otherwise null
 */
fun Url.findArchitecture(): InstallerManifest.Installer.Architecture? {
    // Architectures higher up in the list have a greater priority. For example, we should check for x86_64 before x86.
    val architectures = listOf(
        "x86_64", "x64", "win64", "x86", "x32", "win32", "i386", "i486", "i586", "i686", "386", "486", "586", "686",
        "arm64", "arm", "aarch64", "aarch", "amd64", "neutral"
    )
    val delimiter = "[,\\._-]"
    val archInUrl = "(?<=$delimiter)(${architectures.joinToString("|")})(?=$delimiter)"
        .toRegex(RegexOption.IGNORE_CASE)
        .find(fullPath)
        ?.value
        ?.lowercase()

    return archInUrl?.let { arch ->
        when {
            arch.startsWith("aarch") -> InstallerManifest.Installer.Architecture.ARM.takeIf { arch == "aarch" }
                ?: InstallerManifest.Installer.Architecture.ARM64
            arch == "x86_64" || arch == "win64" || arch == "amd64" -> InstallerManifest.Installer.Architecture.X64
            arch matches Regex("i?[3-6]86") || arch == "x32" || arch == "win32" -> InstallerManifest.Installer.Architecture.X86
            else -> InstallerManifest.Installer.Architecture.valueOfOrNull(arch.uppercase())
        }
    }
}

/**
 * Returns the extension of this Url.
 *
 * The extension is defined as the substring after the last occurrence of the '.' character in the full path of the Url.
 * If there are any non-alphanumeric characters in the extension, they are removed.
 * If no extension is found, the default value "winget-tmp" is returned.
 *
 * @return The extension of this Url, or "winget-tmp" if no extension is found.
 */
val Url.extension get() = fullPath
    .substringAfterLast('.')
    .split("[^A-Za-z0-9]".toRegex())
    .firstOrNull()
    ?: "winget-tmp"

/**
 * Returns the filename of this URL, including the extension, if it has one.
 *
 * The filename is determined by searching the URL's path segments for the last segment that ends with the same
 * extension as the URL's extension, obtained by calling the getExtension() function. If no such segment is found, null
 * is returned.
 *
 * @return the filename of this URL, including the extension, or null if no such filename can be determined
 */
fun Url.getFileName(): String? = pathSegments.findLast { it.endsWith(".$extension") }

/**
 * Returns the filename of this URL without the extension, if it has one.
 *
 * The filename without the extension is determined by calling the getFileName() function to obtain the full filename,
 * then removing the extension by calling the removeSuffix() function with the URL's extension, obtained by calling
 * the getExtension() function. If the getFileName() function returns null, null is returned.
 *
 * @return the filename of this URL without the extension, or null if no such filename can be determined
 */
fun Url.getFileNameWithoutExtension(): String? = getFileName()?.removeSuffix(".$extension")

/**
 * Determines the installation scope of the installer manifest at this URL, if any.
 *
 * The installation scope is determined by searching the URL's full path for the strings "user" or "machine",
 * ignoring case. If "user" is found, the installation scope is [InstallerManifest.Scope.User].
 * If "machine" is found, the installation scope is [InstallerManifest.Scope.Machine].
 * If neither string is found, null is returned.
 *
 * @return the installation scope of the installer manifest at this URL, or null if no scope can be determined
 */
fun Url.findScope(): InstallerManifest.Scope? {
    return when {
        fullPath.contains(other = "user", ignoreCase = true) -> InstallerManifest.Scope.User
        fullPath.contains(other = "machine", ignoreCase = true) -> InstallerManifest.Scope.Machine
        else -> null
    }
}

/**
 * Retrieves the final URL after all redirects have been followed, given a [client] of [HttpClient].
 *
 * @param client a [HttpClient] instance to make the request.
 * @return the final URL after all redirects have been followed.
 */
suspend fun Url.getRedirectedUrl(client: HttpClient = Http.client): Url {
    client.config { followRedirects = false }.use { noRedirectClient ->
        var redirectedInstallerUrl: Url = this
        var response: HttpResponse? = noRedirectClient.head(this)

        var status: HttpStatusCode? = response?.status
        var location: String? = response?.run { headers[HttpHeaders.Location] }
        var redirectCount = 0
        while (status?.isRedirect == true && location != null && redirectCount < 5) {
            redirectedInstallerUrl = Url(location)
            response = noRedirectClient.head(redirectedInstallerUrl)
            status = response.status
            location = response.headers[HttpHeaders.Location]
            redirectCount++
        }
        return redirectedInstallerUrl
    }
}

/**
 * Determines whether this HTTP status code represents a redirect.
 *
 * A status code is considered a redirect if its value is between HttpStatusCode.MultipleChoices.value (300)
 * and HttpStatusCode.PermanentRedirect.value (308), inclusive.
 *
 * @return true if this status code represents a redirect, false otherwise
 */
val HttpStatusCode.isRedirect: Boolean
    get() = value in HttpStatusCode.MultipleChoices.value..HttpStatusCode.PermanentRedirect.value

File: src/url_utils.rs
use crate::installer_manifest::{Architecture, Scope};

pub const VALID_FILE_EXTENSIONS: [&str; 7] = [
    "msix",
    "msi",
    "appx",
    "exe",
    "zip",
    "msixbundle",
    "appxbundle",
];

const DELIMITERS: [char; 6] = [',', '/', '\\', '.', '_', '-'];

const ARCHITECTURES: [(&str, Architecture); 29] = [
    ("x86_64", Architecture::X64),
    ("x64", Architecture::X64),
    ("64-bit", Architecture::X64),
    ("64bit", Architecture::X64),
    ("win64", Architecture::X64),
    ("winx64", Architecture::X64),
    ("ia64", Architecture::X64),
    ("amd64", Architecture::X64),
    ("x86", Architecture::X86),
    ("x32", Architecture::X86),
    ("32-bit", Architecture::X86),
    ("32bit", Architecture::X86),
    ("win32", Architecture::X86),
    ("winx86", Architecture::X86),
    ("ia32", Architecture::X86),
    ("i386", Architecture::X86),
    ("i486", Architecture::X86),
    ("i586", Architecture::X86),
    ("i686", Architecture::X86),
    ("386", Architecture::X86),
    ("486", Architecture::X86),
    ("586", Architecture::X86),
    ("686", Architecture::X86),
    ("arm64", Architecture::Arm64),
    ("aarch64", Architecture::Arm64),
    ("arm", Architecture::Arm),
    ("armv7", Architecture::Arm),
    ("aarch", Architecture::Arm),
    ("neutral", Architecture::Neutral),
];

pub fn find_architecture(url: &str) -> Option<Architecture> {
    // Ignore the casing of the URL
    let url = url.to_lowercase();

    // Check for {delimiter}{architecture}{delimiter}
    for (arch_name, arch) in ARCHITECTURES {
        if url.contains(arch_name) {
            let mut url_chars = url.chars();
            // Get the character before the architecture, consuming the characters before it
            let char_before_arch = url_chars.nth(url.rfind(arch_name).unwrap_or(usize::MAX) - 1);
            // As the characters have been consumed, we can skip by the length of the architecture
            let char_after_arch = url_chars.nth(arch_name.chars().count());
            // If the architecture is surrounded by valid delimiters, the architecture is valid
            if DELIMITERS.contains(&char_before_arch.unwrap_or_default())
                && DELIMITERS.contains(&char_after_arch.unwrap_or_default())
            {
                return Some(arch);
            }
        }
    }

    // If the architecture has not been found, check for {architecture}.{extension}
    let extensions = VALID_FILE_EXTENSIONS
        .into_iter()
        .filter(|extension| url.contains(extension));
    for extension in extensions {
        for (arch_name, arch) in ARCHITECTURES {
            if url.contains(&format!("{arch_name}.{extension}")) {
                return Some(arch);
            }
        }
    }

    None
}

pub fn find_scope(url: &str) -> Option<Scope> {
    match url.to_lowercase() {
        url if url.contains("user") => Some(Scope::User),
        url if url.contains("machine") => Some(Scope::Machine),
        _ => None,
    }
}

#[cfg(test)]
mod tests {
    use crate::installer_manifest::Architecture;
    use crate::url_utils::find_architecture;
    use rstest::rstest;

    #[rstest]
    fn test_x64_architectures(
        #[values("x86_64", "x64", "64-bit", "64bit", "Win64", "Winx64", "ia64", "amd64")]
        architecture: &str,
        #[values(",", "/", "\\", ".", "_", "-", "")] delimiter: &str,
    ) {
        assert_eq!(
            Some(Architecture::X64),
            find_architecture(&format!(
                "https://www.example.com/file{delimiter}{architecture}.exe"
            ))
        );
    }

    #[rstest]
    fn test_x86_architectures(
        #[values(
            "x86", "x32", "32-bit", "32bit", "win32", "winx86", "ia32", "i386", "i486", "i586",
            "i686", "386", "486", "586", "686"
        )]
        architecture: &str,
        #[values(",", "/", "\\", ".", "_", "-", "")] delimiter: &str,
    ) {
        assert_eq!(
            Some(Architecture::X86),
            find_architecture(&format!(
                "https://www.example.com/file{delimiter}{architecture}.exe"
            ))
        );
    }

    #[rstest]
    fn test_arm64_architectures(
        #[values("arm64", "aarch64")] architecture: &str,
        #[values(",", "/", "\\", ".", "_", "-", "")] delimiter: &str,
    ) {
        assert_eq!(
            Some(Architecture::Arm64),
            find_architecture(&format!(
                "https://www.example.com/file{delimiter}{architecture}.exe"
            ))
        );
    }
    #[rstest]
    fn test_arm_architectures(
        #[values("arm", "armv7", "aarch")] architecture: &str,
        #[values(",", "/", "\\", ".", "_", "-", "")] delimiter: &str,
    ) {
        assert_eq!(
            Some(Architecture::Arm),
            find_architecture(&format!(
                "https://www.example.com/file{delimiter}{architecture}.exe"
            )),
        );
    }

    #[test]
    fn test_no_architecture() {
        assert_eq!(None, find_architecture("https://www.example.com/file.exe"));
    }
}
